storage/innobase/handler/ha_innodb.cc

```diff
@@ -18757,22 +18755,25 @@ wsrep_abort_transaction(
 	ut_ad(bf_thd);
 	ut_ad(victim_thd);
 
-	wsrep_thd_kill_LOCK(victim_thd);
-	wsrep_thd_LOCK(victim_thd);
+	/* Caller holds the victim LOCK_thd_data, see ha_abort_transaction()
+	 * in handler.cc. */
+
 	trx_t* victim_trx= thd_to_trx(victim_thd);
-	wsrep_thd_UNLOCK(victim_thd);
 
 	WSREP_DEBUG("abort transaction: BF: %s victim: %s victim conf: %s",
 			wsrep_thd_query(bf_thd),
 			wsrep_thd_query(victim_thd),
 			wsrep_thd_transaction_state_str(victim_thd));
 
 	if (victim_trx) {
-		victim_trx->lock.set_wsrep_victim();
-
-		wsrep_thd_LOCK(victim_thd);
+		victim_trx->mutex_lock();
+		if (victim_trx->state == TRX_STATE_ACTIVE)
+		{
+			victim_trx->lock.set_wsrep_victim();
+		}
+		victim_trx->mutex_unlock();
```

Comment:
This is the point where:
Can victim transaction move from ACTIVE to COMMITTED_IN_MEMORY so that flag is still there ? or it is cleared during commit ?

Reply:
The transaction state is changed to  `TRX_STATE_COMMITTED_IN_MEMORY` in `trx_t::commit_state()`. This is protected by `TMTrxGuard` and happens before the end of `trx_t::commit_in_memory()` where the `lock.was_chose_as_deadlock_victim` is set to false.

If the `lock.was_chosen_as_deadlock_victim` is set just before commit, the commit can still proceed as the flag is not checked in commit code path, and the flag is cleared at the end of `trx_t::commit_in_memory()`. With this fix the flag cannot be set to true after the transaction has passed `trx_t::commit_in_memory()`.

Reply:
TODO: Add comment in code which clarifies this.

Reply:
After code reading I can see that you are correct.

Reply:
There is one more possible race when autocommit non-locking transaction changes its state to `TRX_STATE_NOT_STARTED` in `commit_in_memory()`. Not sure what to do about it...